File: beecrowd/python/estruturas-e-bibliotecas/ex1083.py
operadores = {'^': 6, '*': 5, '/': 5, '+': 4, '-': 4, '>': 3, '<': 3, '=': 3, '#': 3, '.': 2, '|': 1}
operandos = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'

def isValid(expressao):
    stack = []
    isOperador = False
    if expressao[0] in operadores or expressao[-1] in operadores:
        return 'Syntax Error!'
    for index, e in enumerate(expressao): 
        if e in operadores:
            if isOperador:
                return 'Syntax Error!'
            else:
                isOperador = True
        else:
            isOperador = False
        if e == '(':
            stack.append('(')
        elif e == ')':
            if len(stack) == 0:
                return 'Syntax Error!'
            stack.pop()
        elif e not in operadores and e not in operandos:
            return 'Lexical Error!'

    if len(stack) == 0:
        return converterPosfixa(expressao)
    else:
        return 'Syntax Error!'
    
def converterPosfixa(expressao):
    oppilha = []
    listsaida = []
    expressao = list(expressao)
    for e in expressao:
        if e in operandos:
            listsaida.append(e)
        elif e == '(':
            oppilha.append('(')
        elif e == ')':
            item = oppilha.pop()
            while item != '(':
                listsaida.append(item)
                item = oppilha.pop()
        else:
            while (not len(oppilha) == 0) and (oppilha[-1] != '('):
                if operadores[oppilha[-1]] >= operadores[e]:
                    listsaida.append(oppilha.pop())
                else:
                    break
            oppilha.append(e)
    while len(oppilha) != 0:
        listsaida.append(oppilha.pop())
    return "".join(listsaida)       
    
while True:
    try:
        expressao = input()
        print(isValid(expressao))
    except EOFError: 
        break

File: beecrowd/c++/estruturas-e-bibliotecas/ex1083.cpp
#include <iostream>
#include <string>
#include <map>
#include <stack>

using namespace std;

map<char, int> operadores = {{'^', 6}, {'*', 5}, {'/', 5}, {'+', 4}, {'-', 4}, {'>', 3}, {'<', 3}, {'=', 3}, {'#', 3}, {'.', 2}, {'|', 1}};
string operandos = "aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ0123456789";

bool isOperando(char c){
    for (char o : operandos){
        if (c == o){
            return true;
        }
    }
    return false;
}

int isValid(string exp){
    stack<char> parenteses;
    bool isOperador = false;

    if (operadores.count(exp[0])==1 || operadores.count(exp[exp.size()-1])==1){
        return 1;
    }

    for (char c : exp){
        if(operadores.count(c) == 1){
            if (isOperador){
                return 1; //Syntax Error
            } else{
                isOperador = true;
            }
        } else{
            isOperador = false;
        }
        
        if (c=='('){
            parenteses.push('(');
        } else if(c==')'){
            if(parenteses.empty()){
                return 1; //Syntax Error
            }
            parenteses.pop();
        } else if (!isOperando(c) && operadores.count(c) != 1){
            return 2; // Lexical Error
        }
    }
    if (!parenteses.empty()){
        return 1; // Syntax Error
    }
    return 0; //É válido
}

string converterPosfixa(string exp){
    stack<char> oper;
    string posf = "";

    for (char c : exp){
        if (isOperando(c)){
            posf += c;
        } else if(c == '('){
            oper.push('(');
        } else if(c == ')'){
            char item = oper.top();
            oper.pop();
            while (item != '('){
                posf += item;
                item = oper.top();
                oper.pop();
            }
        } else{
            while (!oper.empty() && oper.top()!='('){
                if (operadores[oper.top()] >= operadores[c]){
                    posf += oper.top();
                    oper.pop();
                } else{
                    break;
                }
            }
            oper.push(c);
        }
    }

    while (!oper.empty()){
        posf += oper.top();
        oper.pop();
    }

    return posf;
}

int main(){
    string exp;
    string posf;

    while (getline(cin, exp)){
        int valid = isValid(exp);
        if(valid == 1){
            cout<<"Syntax Error!"<<endl;
        } else if(valid == 2){
            cout<<"Lexical Error!"<<endl;
        } else{
            cout<<converterPosfixa(exp)<<endl;
        }
    }


    return 0;
}